accepted/2023/sign-signature-provider-plugins.md

```diff
@@ -0,0 +1,524 @@
+               TargetPath="$(ProjectDepsFileName)"
+               FinalOutputPath="$(ProjectDepsFilePath)" />
+       </ItemGroup>
+   </Target>
+   ```
+
+#### <a name="plugin-json-file"></a>The `plugin.json` file
+
+Sign CLI needs to load and execute plugins.  The general problem is that Sign CLI needs to know which assemblies in a plugin to load, which types to instantiate, how to initialize those objects, and so forth.  To simplify matters, plugins will embed this information in a `plugin.json` JSON file in their package's root directory.  The file should include the following properties:
+
+* `name`: The plugin's command name (e.g.:  `azure-key-vault` in `sign code azure-key-vault`).
+* `description`: The plugin's command descripton, to be displayed in command help.
+* `entryPoints`: Information for plugin instantiation.  The value is an object of key-value pairs:
+  * `<tfm>`: The target framework moniker (TFM) (e.g.:  `net6.0`, `net7.0`, `net8.0`) that the entry point targets.
+    * `filePath`: The full file path within the package, relative to the package's root directory, for the assembly which contains a public implementation of an interface defined in the interfaces package.
+      * The path must be in its simplest form, without `..` or `.` directories.
+      * The directory separator must be `/`.
+      * The path must not have a leading slash `/`.
+      * The path must be case-sensitive.
+    * `implementationTypeName`: The fully qualified type name of the public type implementing the public interface defined in the interfaces package.
+    * `interfaceTypeName`: The fully qualified type name of the public type in the interfaces package that `implementationTypeName` implements.
+* `parameters`: The plugin's command options.
+  * `name`: The option name.  Not displayed.
+  * `description`: The option's description.  Displayed in command help.
+  * `aliases`: The option's names.  Should include both long and short forms, (e.g.:  `--azure-key-vault-managed-identity` and `-kvm`, respectively).  A user will type these option names.
+  * `dataType`: The option's data type.  The default is `String`, and all `String` options require an explicit value.  When this value is `Boolean`, it provides a hint to Sign CLI to create a switch option that supports both explicit value (e.g.:  `--global true`) and implicit value syntaxes (e.g.:  `--global`).
+  * `defaultValue`: The option's default value.  Only used if `isRequired` is `false`; otherwise, this is ignored.
+  * `isRequired`: Whether the option is required or not.
+
+Parameter | Required | JSON Type | Default | Possible Values
+-- | -- | -- | -- | --
+`name` | yes | string | N/A | N/A
+`description` | yes | string | N/A | N/A
+`aliases` | yes | array of strings | N/A | N/A
+`dataType` | no | string | `Text` | `Text`, `Boolean`
+`defaultValue` | no | string | N/A | N/A
+`isRequired` | no | `true`/`false` | `false` | `true`, `false`
+
+Example:
+
+```JSON
+{
```

Comment:
Do we need some kind of indicator that it is a plugin for Sign CLI specifically?

Reply:
Alternatively, NuGet package type mechanism could be used for that.

Reply:
@agr, if we move to a more generic .NET tool plugin pattern, then probably yes.